test/config/config_settings_spec.py

```diff
@@ -35,31 +36,31 @@
             assert results == invalid_examples
 
         with it("should set the empty textboxes to their default values"):
-            config = ConfigSettings()
-
-            config.export_note_sort_methods = [""]
-            config.snapshot_path = ""
+            config = ConfigSettings({
+                "export_note_sort_methods": [""],
+                "snapshot_path": ""
+            })
 
             config.handle_empty_textboxes()
 
             assert config.export_note_sort_methods == config.Properties.EXPORT_NOTE_SORT_METHODS.value.default_value
             assert config.snapshot_path == config.Properties.SNAPSHOT_PATH.value.default_value
 
-        # with it("tries to save"):
-        #     config = ConfigSettings()
-        #
-        #     config.snapshot_path = "testing"
-        #     config.automated_snapshot = True
-        #     config.snapshot_root_decks = ["TestDeck1", "Other"]
-        #     config.export_notes_reverse_order = True
-        #     config.export_note_sort_methods = ["notemodel", "guid"]
-        #
-        #     config.save()
-        #
-        #     print(config._config["snapshot_path"])
-        #
-        #     assert config._config["snapshot_path"] == config.snapshot_path
-        #     assert config._config["automated_snapshot"] == config.automated_snapshot
-        #     assert config._config["snapshot_root_decks"] == config.snapshot_root_decks
-        #     assert config._config["export_notes_reverse_order"] == config.export_notes_reverse_order
-        #     assert config._config["export_note_sort_methods"] == config.export_note_sort_methods
+        with it("tries to save"):
+            config = ConfigSettings({
+                "snapshot_path": "testing",
+                "automated_snapshot": True,
+                "snapshot_root_decks": ["TestDeck1", "Other"],
+                "export_notes_reverse_order": True,
+                "export_note_sort_methods": ["notemodel", "guid"]
+            })
+
+            config.save()
+
+            print(config._config["snapshot_path"])
+
+            assert config._config["snapshot_path"] == config.snapshot_path
+            assert config._config["automated_snapshot"] == config.automated_snapshot
+            assert config._config["snapshot_root_decks"] == config.snapshot_root_decks
+            assert config._config["export_notes_reverse_order"] == config.export_notes_reverse_order
+            assert config._config["export_note_sort_methods"] == config.export_note_sort_methods
```

Comment:
I think a better (and also easier to write) test would be to verify that this dict gets passed in to the anki mock (as then you can reuse your dict definition above and the code would go the full circle from init to save)

Reply:
Though I remember you've had trouble interacting with that mock, so if you can't figure it out this should be ok as well

Reply:
Ah, on looking at it again now - I think it'd be a good idea to pass in the main window (or even just addonManager probably) as a parameter to constructor, with None as default value (similar to what we do for dict now).

That way you'd be able to pass in the mock you want from here, and do the testing properly

Reply:
Actually I just realised from your comment, that this does not test the saving. I am checking the ConfigSettings' _config dictionary, that those values are set. Instead I should have been checking if the attributes of ConfigSettings get set based on said dictionary. Changed to your suggestion, and testing properly now:
```

        with it("tries to save"):
            settings = {
                "snapshot_path": "testing",
                "automated_snapshot": True,
                "snapshot_root_decks": ["TestDeck1", "Other"],
                "export_notes_reverse_order": True,
                "export_note_sort_methods": ["notemodel", "guid"]
            }

            config = ConfigSettings(settings)

            config.save()

            assert config._config == settings

            for key in settings:
                assert settings[key] == getattr(config, key)

```

All working :+1: thanks for the comment

Reply:
still may want to verify that save on the actual anki is called.

Reply:
doing getattr actually tests loading and not saving. in save you're copying things over to that map (so it's helpful to assert that it looks like what you'd expect) and then call anki to save the map to file

Reply:
Ahh I see, I got myself muddled up. This test only worked because the load_values function is called in the constructor of ConfigSettings. Changed to two separate tests, which test both the loading and saving :+1: 

```
        with it("tries to load to initial values in the constructor"):
            settings = {
                "snapshot_path": "testing",
                "automated_snapshot": True,
                "snapshot_root_decks": ["TestDeck1", "Other"],
                "export_notes_reverse_order": True,
                "export_note_sort_methods": ["notemodel", "guid"]
            }

            config = ConfigSettings(init_values=settings)

            assert config._config == settings

            for key in settings:
                assert settings[key] == getattr(config, key)

        with it("tries to save new values to config dictionary"):
            addon_manager_mock = MagicMock()

            old_settings = {
                "snapshot_path": "",
                "automated_snapshot": False,
                "snapshot_root_decks": [],
                "export_notes_reverse_order": False,
                "export_note_sort_methods": []
            }

            new_settings = {
                "snapshot_path": "testing",
                "automated_snapshot": True,
                "snapshot_root_decks": ["TestDeck1", "Other"],
                "export_notes_reverse_order": True,
                "export_note_sort_methods": ["notemodel", "guid"]
            }

            config = ConfigSettings(addon_manager=addon_manager_mock, init_values=old_settings)

            for key in new_settings:
                setattr(config, key, new_settings[key])

            config.save()

            assert config._config == new_settings
            addon_manager_mock.writeConfig.assert_called_once()

            for key in new_settings:
                assert new_settings[key] == config._config[key]

```